DirectProgramming/DPC++FPGA/ReferenceDesigns/db/src/dbdata.hpp

```diff
@@ -0,0 +1,214 @@
+constexpr int kCustomerTableSize = kSF * 150000;
+// LINEITEM table is not a strict multiple of kSF
+constexpr int LineItemTableSizeFnc() {
+  if (kSF == 0.01f) {
+    return 60175;
+  } else if (kSF == 1.0f) {
+    return 6001215;
+  } else {
+    return 0;  // error, should be caught by kSF static_assert
+  }
+}
+constexpr int kLineItemTableSize = LineItemTableSizeFnc();
+constexpr int kNationTableSize = 25;
+constexpr int kRegionTableSize = 5;
+
+constexpr int kLineStatusSize = 2;
+constexpr int kReturnFlagSize = 3;
+constexpr int kQuery1OutSize = kReturnFlagSize * kLineStatusSize;
+
+// helpers
+DBDate DateFromString(std::string& date_str);
+int ShipmodeStrToInt(std::string& shipmode_str);
+
+// LINEITEM table
+struct LineItemTable {
+  std::vector<DBIdentifier> orderkey;
+  std::vector<DBIdentifier> partkey;
+  std::vector<DBIdentifier> suppkey;
+  std::vector<DBInt> linenumber;
+  std::vector<DBDecimal> quantity;
+  std::vector<DBDecimal> extendedprice;
+  std::vector<DBDecimal> discount;
+  std::vector<DBDecimal> tax;
+  std::vector<char> returnflag;
+  std::vector<char> linestatus;
+  std::vector<DBDate> shipdate;
+  std::vector<DBDate> commitdate;
+  std::vector<DBDate> receiptdate;
+  std::vector<char> shipinstruct;
+  std::vector<int> shipmode;
+  std::vector<char> comment;
+
+  size_t rows;
+};
+
+// ORDERS table
+struct OrdersTable {
+  std::vector<DBIdentifier> orderkey;
+  std::vector<DBIdentifier> custkey;
+  std::vector<char> orderstatus;
+  std::vector<DBDecimal> totalprice;
+  std::vector<DBDate> orderdate;
+  std::vector<int> orderpriority;
+  std::vector<char> clerk;
+  std::vector<int> shippriority;
+  std::vector<char> comment;
+
+  size_t rows;
+};
+
+// PARTS table
+struct PartsTable {
+  std::vector<DBIdentifier> partkey;
+  std::vector<char> name;
+  std::vector<char> mfgr;
+  std::vector<char> brand;
+  std::vector<char> type;
+  std::vector<int> size;
+  std::vector<char> container;
+  std::vector<DBDecimal> retailprice;
+  std::vector<char> comment;
+
+  size_t rows;
+};
+
+// SUPPLIER table
+struct SupplierTable {
+  std::vector<DBIdentifier> suppkey;
+  std::vector<char> name;
+  std::vector<char> address;
+  std::vector<unsigned char> nationkey;
+  std::vector<char> phone;
+  std::vector<DBDecimal> acctbal;
+  std::vector<char> comment;
+
+  size_t rows;
+};
+
+// PARTSUPP table
+struct PartSupplierTable {
+  std::vector<DBIdentifier> partkey;
+  std::vector<DBIdentifier> suppkey;
+  std::vector<int> availqty;
+  std::vector<DBDecimal> supplycost;
+  std::vector<char> comment;
+
+  size_t rows;
+};
+
+// NATION table
+struct NationTable {
+  std::vector<DBIdentifier> nationkey;
+  std::vector<char> name;
+  std::vector<DBIdentifier> regionkey;
+  std::vector<char> comment;
+
+  std::unordered_map<std::string, unsigned char> name_key_map;
+  std::array<std::string, kNationTableSize> key_name_map;
+  size_t rows;
+};
+
+// the database
+struct Database {
+  LineItemTable l_;
```

Comment:
Only private member variables should be with an underscore. (struct is all public).